src/jvm/main/org/jetbrains/kotlinx/lincheck/util/Logger.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2025 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck.util
+
+import java.io.*
+import java.text.SimpleDateFormat
+import java.util.*
+
+object LincheckLogger {
+    private val LINE_SEPARATOR = System.lineSeparator()
+    private val formatter: ThreadLocal<SimpleDateFormat> = object : ThreadLocal<SimpleDateFormat>() {
+        override fun initialValue(): SimpleDateFormat {
+            return SimpleDateFormat("dd.MM.yyyy HH:mm:ss")
+        }
+    }
+
+    private val logFilename = System.getProperty("lincheck.log.file") ?: null
+    private val logFile = if (logFilename != null) File(logFilename) else null
+    private var logWriter: Writer? = null
+
+    init {
+        println("property: lincheck.log.file = " + System.getProperty("lincheck.log.file"))
+        logFile?.let {
+            initFile(it)
+            logWriter = BufferedWriter(FileWriter(it))
+            println("Lincheck log file could be found at: ${it.absolutePath}")
+        }
+    }
+
+    /**
+     * Logs will be printed to log file if it exists.
+     */
+    fun log(s: String) {
+        logFile?.let {
+            write(s, logWriter!!)
+        }
+    }
+
+    /**
+     * Errors will be printed to log file if it is set.
+     * However, they will always be printed to the `System.err`.
+     */
+    fun error(m: String, e: Throwable) {
```

Comment:
Let's create logging methods for each logging level:

```kotlin
fun error(lazyMessage: () -> String)
fun warning(lazyMessage: () -> String)
fun info(lazyMessage: () -> String)
fun debug(lazyMessage: () -> String)
```

Then for `error` and `warning` you can also versions that take exception as an argument:
```kotlin
fun error(throwable: Threowable) {
    error { throwable.toString() }
}
```